### PR TITLE
Add AWS S3 i/o

### PR DESCRIPTION
**Pull request by Laksh1997**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## What does this PR do?
Allows user to save and load checkpoints to S3 paths (e.g "s3://my-bucket/my-folder/my-checkpoint.ckpt")

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/io.py [hunk 1]
```diff
@@ -8,4 +8,47 @@ def load(path_or_url: str, map_location=None):
     if parsed.scheme == '':
         # local file
         return torch.load(path_or_url, map_location=map_location)
+    elif parsed.scheme == 's3':
+        return load_s3_checkpoint(path_or_url, map_location=map_location)
     return torch.hub.load_state_dict_from_url(path_or_url, map_location=map_location)
+
+
+def load_s3_checkpoint(checkpoint_path, map_location, **pickle_load_args):
+    from torch.serialization import _legacy_load
```
Comment (justusschock) [inline pytorch_lightning/utilities/io.py:17]: why is `_legacy_load` required here? I think we should aim for `torch.load` here

File: pytorch_lightning/utilities/io.py [hunk 2]
```diff
@@ -8,4 +8,47 @@ def load(path_or_url: str, map_location=None):
...
+    if 'encoding' not in pickle_load_args.keys():
+        pickle_load_args['encoding'] = 'utf-8'
+    fs = s3fs.S3FileSystem()
+    with fs.open(checkpoint_path, "rb") as f:
+        checkpoint = _legacy_load(f, map_location, pickle, **pickle_load_args)
+    return checkpoint
+
+
+def save_s3_checkpoint(checkpoint, checkpoint_path):
+    from torch.serialization import _legacy_save
```
Comment (justusschock) [inline pytorch_lightning/utilities/io.py:38]: same as above. IMO we should aim for `torch.save`

Comment (Laksh1997) [reply pytorch_lightning/utilities/io.py:38]: `torch.save` can only save locally. We could use `boto3` to upload locally saved files. Thoughts?

File: pytorch_lightning/utilities/io.py (hunk 1 above)
Comment (Laksh1997) [reply pytorch_lightning/utilities/io.py:17]: We could use `boto3` to download an S3 file, then use `torch.load`. Thoughts?

**Comment by williamFalcon** [comment]
@Borda this has failing tests

**Comment by Laksh1997** [comment]
Thoughts on `_legacy_save` and direct write to S3 vs normal `torch.save` to local, then `boto3` upload to S3?

**Comment by Borda** [comment]
> @Borda this has failing tests

I am fixing master here #2176

> Thoughts on `_legacy_save` and direct write to S3 vs normal `torch.save` to local, then `boto3` upload to S3?

depends how long the legacy func will be available in PyTorch
@Laksh1997 mind add a test for these new functions?

**Comment by Laksh1997** [comment]
I'm thinking about how to add tests. 

One is to just save and load a file actually to an open-access S3 bucket.

However, another way is to use the `moto` framework. This would require using `boto3` for upload however. 

Thinking what the best solution is, any help appreciated.

**Comment by awaelchli** [comment]
> Thoughts on `_legacy_save` and direct write to S3 vs normal `torch.save` to local, then `boto3` upload to S3?

anything that has "legacy" in it's name is a big alarm sign for me hehe :) 
aren't we supposed to use TorchServe to upload stuff to the amazon servers?

**Comment by Laksh1997** [comment]
@awaelchli I agree that `_legacy_save` raises alarms.

Shall we do this then?

- User passes in checkpointing filepath, eg `s3://my-bucket/my-key/checkpoint.ckpt`

- Initially, we save it locally normally with `torch.save` with filepath `my-key/checkpoint.ckpt`

- We then upload the file to the s3 path as specified using `boto3`

The only complication with the above is the following: If we have model checkpointing configured to only save the best 5 models, then we'll end up saving many more to S3 because it will upload every time a new model entry occurs. So to get around this, we would also have to accordingly delete files on S3.

Thoughts?

**Comment by Laksh1997** [comment]
Thoughts on deleting the cloud checkpoint if it's not in the top K models anymore?

We would need to edit `_del_model` in `model_checkpoint.py` on line 158

**Comment by Laksh1997** [comment]
Alright moving everything over to boto3. Will make tests easier that way with moto.

**Comment by f4hy** [comment]
This seems related to #2164 which would support s3/gcs/hdfs.

**Comment by Laksh1997** [comment]
@f4hy Yeah it seems it. What shall we do?

**Comment by f4hy** [comment]
> @f4hy Yeah it seems it. What shall we do?

So I need support for writing all logs and such to a remote path, not just the checkpoints. And I need support for paths other than just s3.

So that's why I set up #2164 to use `gfile`. This problem has been solved elsewhere, so we should just use some other lib. Gfile is great, but it requires `pip install tensorflow` to get full support. Tensorboard has a pruned down implementation that uses the full tf gfile if present otherwise gives just local and s3 support so hopefully that is acceptable and requires no extra dependency.

I think its a mistake to do this in a way that just implements s3. Then we would need to implement the same thing again for gcs or hdfs or others. If the gfile solution is not acceptable I hope we can find some other lib that would do this for us.

**Comment by Laksh1997** [comment]
Hmmm, I agree @f4hy. Let's think of a unified way to do this long term.

I feel however that because S3 is AWS, the cleanest solution to get S3 support would be to use `boto3`. Also, one gets a nice testing framework as well (`moto`)

@Borda @williamFalcon I've just redone the PR to now work using `boto3`. Will be adding `moto` tests soon. Any comments please let me know.

**Comment by yukw777** [comment]
I mentioned this in https://github.com/PyTorchLightning/pytorch-lightning/issues/1532, but https://github.com/RaRe-Technologies/smart_open is really nice. It provides a file-like interface for all kinds of transports. I highly recommend using it for this PR!

**Comment by f4hy** [comment]
> I mentioned this in #1532, but https://github.com/RaRe-Technologies/smart_open is really nice. It provides a file-like interface for all kinds of transports. I highly recommend using it for this PR!

How is this better than the same gfile interface provided by tensorboard. Again #2164  but we dont need to add a new dependency.

**Comment by yukw777** [comment]
@f4hy `smart_open` supports a lot more protocols than the gfile interface with a clean, unified interface with nifty extra features (not that we necessarily need these new features), so we don't need to keep writing new code to support other protocols than s3, gcs, hdfs. I personally think the benefit of introducing this new dependency is worth the effort (i'm currently using it for work, and it's been working quite nicely for us.), but ultimately, if the core team decides to go with another route, I'm fine with that too.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -95,7 +96,7 @@ class ModelCheckpoint(Callback):
 
     def __init__(self, filepath: Optional[str] = None, monitor: str = 'val_loss', verbose: bool = False,
                  save_last: bool = False, save_top_k: int = 1, save_weights_only: bool = False,
-                 mode: str = 'auto', period: int = 1, prefix: str = ''):
+                 mode: str = 'auto', period: int = 1, prefix: str = '', remove_non_top_k_s3_files: bool = True):
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:99]: here I would keep it synced so I would drop `remove_non_top_k_s3_files`

**Comment by Borda** [comment]
@Laksh1997 can we finish this one? :]

**Comment by f4hy** [comment]
I believe #2164 resolved this as well. After the first in #2894 I am able to train and have all outputs go to AWS s3.

**Comment by Borda** [comment]
cool, so let's close this one for now and reopen if needed :]

**Pull request closed without merging**